Add Media Capture Elements (<usermedia>, <camera>, <microphone>) and Explainer (#168)
w3c/mediacapture-extensions: media-capture-elements-explainer.md

# Media Capture Elements: `<usermedia>`, `<camera>`, and `<microphone>`

## Authors:

- Daniel Vogelheim (Google LLC)
- Minh Le (Google LLC)
- Thomas Nguyen (Google LLC)

## Introduction

The web permission model is shifting from passive permission control to active Capability Elements. Building on the foundations of the Page-Embedded Permission Control (PEPC) initiative, these specialized Capability elements evolve the user experience from a simple "Allow/Deny" toggle into a direct controller for web capabilities.

Rather than serving as a static status indicator, a Capability Element acts as a functional action broker. It streamlines the entire lifecycle: establishing a trusted signal of intent, managing the permission flow, and delivering the resulting data or media stream directly to the application. This shift reduces developer friction, eliminates "permission holes," and provides users with a consistent, browser-controlled interface for hardware like cameras and microphones.

> This document proposes a suite of targeted Media Capture Elements (`<usermedia>`, `<camera>`, and `<microphone>`). They move beyond acting merely as permission gatekeepers to serving as data mediators: handling consent, fetching the `MediaStream`, delivering it to the site, and persisting as a stateful, trusted UI control for toggling media tracks.

Reviewer: Order consistency
```suggestion
This document proposes a suite of targeted Media Capture Elements (`<camera>`, `<microphone>`, and `<usermedia>`). They move beyond acting merely as permission gatekeepers to serving as data mediators: handling consent, fetching the `MediaStream`, delivering it to the site, and persisting as a stateful, trusted UI control for toggling media tracks.
```

## Goals

> *   Data Mediator and Action Broker: The elements handle the entire acquisition flow: obtaining user consent, executing the underlying `getUserMedia` request, and exposing the resulting `MediaStream` directly, often eliminating the need for separate JavaScript API boilerplate.

Reviewer: ```suggestion
*   Data Mediator and Action Broker: The elements handle the entire acquisition flow: obtaining user consent, executing the underlying `getUserMedia` request, and exposing the resulting video and/or audio `MediaStreamTrack`s directly, often eliminating the need for separate JavaScript API boilerplate.
```

> *   Stateful Media Capture Control: Elements act as integrated UI components. Following a permission grant, the element transitions from a "request" state to a "control" state. For `<camera>` and `<microphone>`, this provides a browser-enforced, trusted UI toggle for muting and unmuting the stream. All elements act as strict monitors of their assigned track's states, natively reverting to the "request" state if the underlying tracks are fully terminated.

Reviewer: I like the _request_ and _control_ states as explanation, but they're not literal are they? Also, we want to avoid confusion with the muted/unmuted state (which in MediaCapture-main is reserved for UA mute).

```suggestion
*   Stateful Media Capture Control: Elements act as integrated UI components. Following a permission grant, the element transitions from a _request_ state to a _control_ state. For `<camera>` and `<microphone>`, this provides a browser-enforced, UI toggle for disabling and enabling the underlying track. All elements reflect their track's enabled state, and revert to the _request_ state once the track has ended.
```

> *   Reduced Friction & Trusted Intent: Establish a "trusted signal of intent" driven by explicit User Gestures on a browser-controlled element, allowing developers to offer users an option to use the grant and use the capability in context.

Reviewer: ```suggestion
*   Reduced Friction & Trusted Intent: Establish a "trusted signal of intent" driven by explicit User Gestures on a browser-controlled element, allowing developers to offer users the option to use the capability in context without having to deal with permission.
```

> *   Seamless In-Page Recovery: Similar to and inheriting concepts from the `<permission>` element, this resolves the "permission hole" by providing an in-page flow to re-enable access if the user previously denied it, avoiding the need for users to navigate complex, UA-specific browser settings.

Reviewer: s/denied/blocked/ is more implementation agnostic

```suggestion
*   Seamless In-Page Recovery: Similar to and inheriting concepts from the `<permission>` element, this resolves the "permission hole" by providing an in-page flow to re-enable access if the user previously blocked it in the user agent, avoiding the need for users to navigate complex, UA-specific browser settings.
```

*   Specific vs. Generic Tags: Replace the generic `<permission>` element with targeted, semantic elements (`<camera>`, `<microphone>`, `<usermedia>`). This allows for UA-tailored UI, specific IDL attributes, and distinct security/privacy models tailored to the specific hardware media capture capability.

## Non-goals

> *   Replacing `getUserMedia`: The `mediaDevices.getUserMedia` API remains the primary programmatic primitive. Media Capture Elements are designed to complement it by handling common UI-driven flows. In the future, we may add a parameter to indicate whether the `getUserMedia` request originated from a programmatic JS call or directly from a media capture element.

Reviewer: A bit confusing to mention future feature ideas in the non-goals section (I'm also unsure why we would add that). Maybe take it out?

I'd like to highlight somewhere that the permissions API is tied to the old (`getUserMedia`) model. How about:

```suggestion
*   Replacing `getUserMedia`: The `mediaDevices.getUserMedia` API remains the primary programmatic primitive. Media Capture Elements are designed to complement it by handling common UI-driven flows.
*   The `permissions` API will remain to support the use of `getUserMedia` alongside the new elements.
```

*   Complete UI Customization: The shadow DOM/internal UI of these elements is strictly controlled by the UA to prevent clickjacking and ensure a trusted state representation. Complete visual overriding by the site is a non-goal for security reasons.

## The `<camera>` HTML element

> The `<camera>` element provides a dedicated UI for requesting and controlling access to a camera's video track. It renders as a camera text button or icon. After a successful grant, the element transitions to a trusted UI toggle for muting and unmuting the video stream.

Reviewer: ```suggestion
The `<camera>` element provides a dedicated UI for requesting and controlling access to a camera's video track. It renders as a camera text button or icon. After a successful grant, the element transitions to a trusted UI toggle for disabling and enabling the video track.
```

### WebIDL

```webidl
[Exposed=Window]
interface HTMLCameraElement : HTMLElement {
  [HTMLConstructor] constructor();

  readonly attribute MediaStreamTrack? track;
  undefined setConstraints(optional MediaTrackConstraintSet constraints = {});
  readonly attribute any error;

  attribute boolean autostart;

  attribute EventHandler onstream;
  attribute EventHandler ontrackchange;
};
HTMLCameraElement includes ActivationBlockersMixin;
HTMLCameraElement includes PowerfulFeatureObserver;

> ```

Reviewer: I try to discourage WebIDL in explainers. In my view, it belongs in specs (whose audience is implementers).

This also avoids duplication and information getting out of sync.

### Attributes and Properties

*   `track`: A read-only property that holds the associated `MediaStreamTrack` object (video). It is populated automatically upon successful user interaction.
*   `autostart`: A boolean attribute. If present, the UA will attempt to start the stream automatically as the element is added to the DOM and rendered (provided permissions are already granted).
*   `error`: A read-only property that holds a `DOMException` or `null`. Populated if the stream acquisition fails.
*   `onstream`: An `EventHandler` that fires when a `getUserMedia` attempt finishes (either populating `track` or `error`).

> *   `ontrackchange`: An `EventHandler` that fires when the associated track's active/muted/ended state changes.

Reviewer: Can we use JS examples instead to show these in use?

I find using examples helps showcase how use cases are solved for a wider audience unfamiliar with the domain.

See [WebTransport](https://github.com/w3c/webtransport/blob/main/explainer.md) for an example, and let me know if I can help.

> *   `isValid`, `invalidReason`, `onvalidationstatuschange`: Provided by the `ActivationBlockersMixin` to protect against programmatic activation and clickjacking.

Reviewer: No longer in the spec

> *   `onpromptaction`, `onpromptdismiss`: Provided by the `PowerfulFeatureObserver` to monitor permission prompts.

Reviewer: Hopefully not in the spec soon

### Constraints Configuration

Developers can specify a `MediaTrackConstraintSet` to dictate the parameters of the underlying `getUserMedia` call. This is done imperatively via the `setConstraints()` method. If constraints are not set before the user interacts with the element, an empty default (`{}`) is used.

> **Constraint Filtering**: The UA applies a "constraint filter" before executing the `getUserMedia` call. Because the element's `setConstraints()` takes a `MediaTrackConstraintSet` directly (which contains no `audio` or `video` keys), the UA constructs a `MediaStreamConstraints` object where `audio` is forced to `false`, and `video` is set to the provided `MediaTrackConstraintSet`. Additionally, any required constraints (like `exact`, `min`, or `max`) are stripped to prevent the element from failing silently with an `OverconstrainedError` when the user interacts with it.

Reviewer: This seems heavy on implementation detail. Can we describe the net effect instead? The below example already showing `camera.setConstraints({width, height})` I think sufficiently highlights differences from user media.

That, and a note that advanced and required constraints are ignored, might suffice.

### Example

```html
<!-- A camera element that will be configured via JavaScript -->
<camera id="my-camera"></camera>
<video id="camera-playback" autoplay playsinline></video>

<script>
  const cameraEl = document.getElementById('my-camera');
  const videoEl = document.getElementById('camera-playback');

  // The developer configures high-resolution video constraints for the camera element
  cameraEl.setConstraints({

>     width: { ideal: 1920 }, height: { ideal: 1080 }

Reviewer: Please simplify constraints
```suggestion
    width: 1920, height: 1080
```
